src/level_cache.h

```diff
@@ -0,0 +1,126 @@
+        //      or terrain doesn't have "NO_FLOOR" flag
+        // false otherwise
+        // i.e. true == has floor
+        bool floor_cache[MAPSIZE_X][MAPSIZE_Y];
+
+        // stores cached transparency of the tiles
+        // units: "transparency" (see LIGHT_TRANSPARENCY_OPEN_AIR)
+        float transparency_cache[MAPSIZE_X][MAPSIZE_Y];
+
+        // materialized  (transparency_cache[i][j] > LIGHT_TRANSPARENCY_SOLID)
+        // doesn't consider fields (i.e. if tile is covered in thick smoke, it's still
+        // considered transparent for the purpuses of this cache)
+        // true, if tile is not opaque
+        std::array<std::bitset<MAPSIZE_Y>, MAPSIZE_X> transparent_cache_wo_fields;
+
+        // stores "adjusted transparency" of the tiles
+        // initial values derived from transparency_cache, uses same units
+        // examples of adjustment: changed transparency on player's tile and special case for crouching
+        float vision_transparency_cache[MAPSIZE_X][MAPSIZE_Y];
+
+        // stores "visibility" of the tiles to the player
+        // values range from 1 (fully visible to player) to 0 (not visible)
+        float seen_cache[MAPSIZE_X][MAPSIZE_Y];
+
+        // same as `seen_cache` (same units) but contains values for cameras and mirrors
+        // effective "visibility_cache" is calculated as "max(seen_cache, camera_cache)"
+        float camera_cache[MAPSIZE_X][MAPSIZE_Y];
+
+        // reachability caches
+        // Note: indirection here is introduced, because caches are quite large:
+        // at least (MAPSIZE_X * MAPSIZE_Y) * 4 bytes (≈69,696 bytes) each
+        // so having them directly as part of the level_cache interferes with
+        // CPU cache coherency of level_cache
+        cata::value_ptr<reachability_cache_horizontal>r_hor_cache =
+            cata::make_value<reachability_cache_horizontal>();
+        cata::value_ptr<reachability_cache_vertical> r_up_cache =
+            cata::make_value<reachability_cache_vertical>();
+
+        // stores resulting apparent brightness to player, calculated by map::apparent_light_at
+        lit_level visibility_cache[MAPSIZE_X][MAPSIZE_Y];
+        std::bitset<MAPSIZE_X *MAPSIZE_Y> map_memory_seen_cache;
+        std::bitset<MAPSIZE *MAPSIZE> field_cache;
+
+        std::set<vehicle *> vehicle_list;
+        std::set<vehicle *> zone_vehicles;
+
+        bool get_veh_in_active_range() const;
+        bool get_veh_exists_at( const tripoint &pt ) const;
+        std::pair<vehicle *, int> get_veh_cached_parts( const tripoint &pt ) const;
+
+        void set_veh_in_active_range( bool is_active );
+        void set_veh_exists_at( const tripoint &pt, bool exists_at );
+        void set_veh_cached_parts( const tripoint &pt, vehicle *veh, int part_num );
+        void verify_vehicle_cache();
+        void clear_vehicle_cache();
+        void clear_veh_from_veh_cached_parts( const tripoint &pt, vehicle *veh );
+
+    private:
+        bool veh_cache_dirty = true;
```

Comment:
Maybe add some comments on these boolean flags to help with understanding possible states and intended transitions between them?

Reply:
Good idea, I'll do that.